Use MappingQCache for curved cells if possible to speed up particle sort

gassmoeller opened the pull request:
I think this was suggested to me by one of the deal.II maintainers (maybe Martin?) a while ago, and I had a test branch for a while, but did not see much of a change for most applications so I never pursued it further. However, when looking at a particularly slow particle model in a sphere I realized that that is exactly the application getting the largest benefit from using a `MappingQCache` object instead of a `MappingQ` that we used before. The caching of mapping support points drastically speeds up particle sorting in curved geometries, because often there are dozens of particles in a cell, and in the current implementation we recompute the mapping support points for all of them when recomputing their reference locations. The improvement in 3D may be even larger.
The PR should not change test results.

@kchotalia: Please give this PR a try for your models.

Timing of the attached parameter file with 6 processes in release mode and deal.II dev 58ddbdaeab:

This PR:
```
+----------------------------------------------+------------+------------+
| Total wallclock time elapsed since start     |      92.4s |            |
|                                              |            |            |
| Section                          | no. calls |  wall time | % of total |
+----------------------------------+-----------+------------+------------+
| Assemble Stokes system           |       101 |      2.91s |       3.2% |
| Assemble temperature system      |       101 |      4.12s |       4.5% |
| Build Stokes preconditioner      |       101 |      4.74s |       5.1% |
| Build temperature preconditioner |       101 |     0.454s |      0.49% |
| Create snapshot                  |        10 |      1.45s |       1.6% |
| Initialization                   |         1 |      0.14s |      0.15% |
| Particles: Advect                |       202 |       7.7s |       8.3% |
| Particles: Generate              |         1 |     0.132s |      0.14% |
| Particles: Initialization        |         1 |  0.000119s |         0% |
| Particles: Initialize properties |         1 |    0.0663s |         0% |
| Particles: Interpolate           |       101 |      1.34s |       1.4% |
| Particles: Sort                  |       202 |      4.29s |       4.6% |
| Postprocessing                   |       101 |      22.9s |        25% |
| Refine mesh structure, part 1    |        33 |      3.99s |       4.3% |
| Refine mesh structure, part 2    |        33 |      4.84s |       5.2% |
| Setup dof systems                |        34 |     0.546s |      0.59% |
| Setup initial conditions         |         1 |      0.27s |      0.29% |
| Setup matrices                   |        34 |      1.64s |       1.8% |
| Solve Stokes system              |       101 |      27.7s |        30% |
| Solve temperature system         |       101 |     0.488s |      0.53% |
+----------------------------------+-----------+------------+------------+
```

Before:
```
+----------------------------------------------+------------+------------+
| Total wallclock time elapsed since start     |       161s |            |
|                                              |            |            |
| Section                          | no. calls |  wall time | % of total |
+----------------------------------+-----------+------------+------------+
| Assemble Stokes system           |       101 |      2.82s |       1.8% |
| Assemble temperature system      |       101 |       4.1s |       2.6% |
| Build Stokes preconditioner      |       101 |      4.48s |       2.8% |
| Build temperature preconditioner |       101 |      0.41s |      0.26% |
| Create snapshot                  |        10 |      1.21s |      0.75% |
| Initialization                   |         1 |     0.138s |         0% |
| Particles: Advect                |       202 |      6.56s |       4.1% |
| Particles: Generate              |         1 |     0.461s |      0.29% |
| Particles: Initialization        |         1 |  0.000131s |         0% |
| Particles: Initialize properties |         1 |    0.0386s |         0% |
| Particles: Interpolate           |       101 |      1.41s |      0.88% |
| Particles: Sort                  |       202 |        77s |        48% |
| Postprocessing                   |       101 |      21.9s |        14% |
| Refine mesh structure, part 1    |        33 |      3.82s |       2.4% |
| Refine mesh structure, part 2    |        33 |      4.79s |         3% |
| Setup dof systems                |        34 |     0.505s |      0.31% |
| Setup initial conditions         |         1 |     0.581s |      0.36% |
| Setup matrices                   |        34 |      1.47s |      0.91% |
| Solve Stokes system              |       101 |      25.5s |        16% |
| Solve temperature system         |       101 |     0.439s |      0.27% |
+----------------------------------+-----------+------------+------------+
```

[test.prm.txt](https://github.com/geodynamics/aspect/files/5907220/test.prm.txt)

jdannberg commented:
It looks like some tests fail because the mapping hasn't been initialized yet. 
Otherwise, this is great! Thank you!

gassmoeller commented:
I fixed the forgotten initialization (the one right after constructing the coarse mesh, but before the initial refinement). I still have one minor change in the test `hollow_sphere_gmg_project_q1`, but it only changes the nonlinear residuum of a later iteration, and in fact reduces the nonlinear residual, so I guess we are good with that. I updated the one test result. Ready from my side.
